File: Algorithm/C++/1000+/1466. Reorder Routes to Make All Paths Lead to the City Zero.cc
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>

class Solution {
  public:
    int minReorder(int n, std::vector<std::vector<int>>& connections) {
        std::unordered_map<int, std::unordered_set<int>> direction, neighbor;
        std::unordered_set<int> visited;
        std::queue<int> q;
        int result = 0;
        q.push(0);
        for(auto conn: connections) {
            direction[conn[0]].insert(conn[1]);
            neighbor[conn[0]].insert(conn[1]);
            neighbor[conn[1]].insert(conn[0]);
        }
        while (q.size() > 0) {
            int curr = q.front();
            visited.insert(curr);
            q.pop();
            for (int node: neighbor[curr]) {
                if (visited.count(node) != 0) continue;
                if (direction[node].count(curr) == 0) result++;
                q.push(node);
                
            }
        }
        return result;
    }
};

File: Algorithm/Python/1000+/1466. Reorder Routes to Make All Paths Lead to the City Zero.py
from collections import defaultdict, deque
class Solution(object):
    def minReorder(self, n, connections):
        """
        :type n: int
        :type connections: List[List[int]]
        :rtype: int
        """
        neighbor, direction = defaultdict(set), defaultdict(set)
        queue, visited = deque([0]), set()
        for start, end in connections:
            direction[start].add(end)
            neighbor[start].add(end)
            neighbor[end].add(start)
        
        result = 0
        while len(queue) > 0:
            curr = queue.popleft()
            visited.add(curr)
            for node in neighbor[curr]:
                if node in visited:
                    continue
                if curr not in direction[node]:
                    result += 1
                queue.append(node)
        return result                   